pandas/_libs/parsers.pyx

```diff
@@ -986,10 +986,13 @@ cdef class TextReader:
             col_dtype = None
             if self.dtype is not None:
                 if isinstance(self.dtype, dict):
-                    if name in self.dtype:
+                    try:
                         col_dtype = self.dtype[name]
-                    elif i in self.dtype:
-                        col_dtype = self.dtype[i]
+                    except KeyError:
+                        try:
+                            col_dtype = self.dtype[i]
+                        except KeyError:
+                            pass
```

Comment:
Instead of nesting try/excepts here, would it be cleaner to use `get`?

Reply:
Cleaner - maybe but...
1. If you want to save previous behavior you need
```
col_dtype = self.dtype.get(name, None)  # Assuming that explicit is better than implicit
if col_dtype is None:
    col_dtype = self.dtype.get(i, None)
```
It is +2 assignment and +1 comparison in case of the dtype absence for the column (2 try/except are just moved to underhood according to the [Mapping.get implementation](https://github.com/python/cpython/blob/c723d5191110f99849f7b0944820f6c3cd5f7747/Lib/_collections_abc.py#L657) )
2. `col_dtype = self.dtype[i]` can works with list in future but `get` does not

Reply:
On the other hand, I have found [the following example](https://github.com/pandas-dev/pandas/blob/e7c1ef71d80797cab0bffda8c47948261cffc139/pandas/io/parsers/base_parser.py#L522)